File: 0152_maximum_product_subarray/max_product.py
from typing import List


def max_product(self, nums: List[int]) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    max_so_far = min_so_far = res = nums[0]
    for num in nums[1:]:
        tmp_max = max(num, num * max_so_far, num * min_so_far)
        min_so_far = min(num, num * max_so_far, num * min_so_far)
        max_so_far = tmp_max
        res = max(res, max_so_far)
    return res

File: 0152_maximum_product_subarray/max_product.java
class Solution {
    // Time: O(n)
    // Space: O(1)
    public int maxProduct(int[] nums) {
        int maxSoFar = nums[0], minSoFar = nums[0], res = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int curr = nums[i];
            int tmp = Math.max(curr, Math.max(maxSoFar * curr, minSoFar * curr));
            minSoFar = Math.min(curr, Math.min(maxSoFar * curr, minSoFar * curr));
            maxSoFar = tmp;
            res = Math.max(res, maxSoFar);
        }
        return res;
    }
}